src/script/sigcache.cpp

```diff
@@ -53,7 +53,7 @@ class CSignatureCache
     ComputeEntryECDSA(uint256& entry, const uint256 &hash, const std::vector<unsigned char>& vchSig, const CPubKey& pubkey) const
     {
         CSHA256 hasher = m_salted_hasher_ecdsa;
-        hasher.Write(hash.begin(), 32).Write(&pubkey[0], pubkey.size()).Write(&vchSig[0], vchSig.size()).Finalize(entry.begin());
+        hasher.Write(hash.begin(), 32).Write(&pubkey[0], pubkey.size()).Write(vchSig.data(), vchSig.size()).Finalize(entry.begin());
```

Comment:
Why is this changed here, but the other 55 instances are left as is? Including the one in this line and another one in this file?

See also commit 592404f03

And our developer notes:

```
- Watch out for out-of-bounds vector access. `&vch[vch.size()]` is illegal,
  including `&vch[0]` for an empty vector. Use `vch.data()` and `vch.data() +
  vch.size()` instead.

Reply:
Also, it could make sense to open two pull requests for the changes here, since they seem unrelated? (One of them is also failing the fuzz task, btw)

Reply:
> Why is this changed here, but the other 55 instances are left as is? Including the one in this line and another one in this file?

I didn't change the two instances of `&pubkey[0]` in this file because the [CPubkey subscript operator](https://github.com/bitcoin/bitcoin/blob/19a56d1519fb493c3e1bd5cad55360b6b80fa52b/src/pubkey.h#L114) returns a reference to a [regular array](https://github.com/bitcoin/bitcoin/blob/19a56d1519fb493c3e1bd5cad55360b6b80fa52b/src/pubkey.h#L55) instead of a vector, so no undefined behavior rules are violated.

I didn't change any other instances of `&vch[0]` because this instance specifically was the only one which came up with fuzzing (so far) with `-D_GLIBCXX_DEBUG`.

> Also, it could make sense to open two pull requests for the changes here, since they seem unrelated?

@laanwj suggested I create two commits in 1 PR.

> (One of them is also failing the fuzz task, btw)

It fails because it [instantiates ChaCha20 with a key that is ```17..31``` bytes large](https://github.com/bitcoin/bitcoin/blob/19a56d1519fb493c3e1bd5cad55360b6b80fa52b/src/test/fuzz/crypto_chacha20.cpp#L19-L26) whereas my change requires that it is either 16 or 32 bytes.

I can either relax the condition in ChaCha20 to allow keys ```16..32``` bytes large, or change the fuzzer to pick one of `(16, 32)`, let me know which you prefer.

Reply:
> I didn't change any other instances of `&vch[0]` because this instance specifically was the only one which came up with fuzzing (so far) with `-D_GLIBCXX_DEBUG`.

`-D_GLIBCXX_DEBUG` is great! :)

@guidovranken, you don't happen to have any notes saved on any tweaks you had to make to get a full `-D_GLIBCXX_DEBUG` compile including dependencies for Bitcoin Core? IIRC our Boost dependencies were a bit cumbersome to get compiled with `-D_GLIBCXX_DEBUG` at least historically.

If anyone wants to add a CI job which runs the tests (unit + functional + fuzz) under `-D_GLIBCXX_DEBUG` I'd be more than happy to review such a PR.

> I can either relax the condition in ChaCha20 to allow keys `16..32` bytes large, or change the fuzzer to pick one of `(16, 32)`, […]

Asserting the precondition (like you've already done) and modifying the fuzzer to fulfil said precondition is probably the best option.

Reply:
> tweaks you had to make to get a full -D_GLIBCXX_DEBUG compile including dependencies for Bitcoin Core?
> IIRC our Boost dependencies were a bit cumbersome to get compiled with -D_GLIBCXX_DEBUG at least historically.

Note that depends builds with `DEBUG=1` have always used [`-D_GLIBCXX_DEBUG` && `D_GLIBCXX_DEBUG_PEDANTIC`](https://github.com/bitcoin/bitcoin/blob/master/depends/hosts/linux.mk#L10). If building depends with `DEBUG=1` don't work for some reason, please open an issue.

Reply:
Thanks @guidovranken and @fanquake for suggesting both `-D_GLIBCXX_DEBUG` and `-D_GLIBCXX_DEBUG_PEDANTIC`

Looks like there are more issues than this one. I've reported them here:
* https://github.com/boostorg/date_time/issues/190
* https://github.com/bitcoin/bitcoin/pull/21802

Reply:
Is there a downside splitting this into two pulls? I'd still prefer that to make it easier to review, because both commits can be reviewed independently. Also, it makes a cleaner history and discussion because two unrelated threads aren't intertwined.

Reply:
I went ahead and split the raw data pointer commit into a separate pull (#21817). I hope this is ok. The commit hash needed to be changed, but the author information is fully preserved.